File: src/spirite/file/SaveEngine.java
package spirite.file;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import javax.imageio.ImageIO;
import javax.swing.Timer;

import spirite.MDebug;
import spirite.MDebug.ErrorType;
import spirite.MDebug.WarningType;
import spirite.brains.MasterControl;
import spirite.brains.MasterControl.MWorkspaceObserver;
import spirite.image_data.Animation;
import spirite.image_data.GroupTree;
import spirite.image_data.GroupTree.GroupNode;
import spirite.image_data.GroupTree.Node;
import spirite.image_data.ImageHandle;
import spirite.image_data.ImageWorkspace;
import spirite.image_data.animation_data.FixedFrameAnimation;
import spirite.image_data.animation_data.FixedFrameAnimation.AnimationLayer;
import spirite.image_data.animation_data.FixedFrameAnimation.AnimationLayer.Frame;
import spirite.image_data.animation_data.FixedFrameAnimation.Marker;
import spirite.image_data.layers.Layer;
import spirite.image_data.layers.RigLayer;
import spirite.image_data.layers.RigLayer.Part;
import spirite.image_data.layers.SimpleLayer;

/***
 * SaveEngine is a static container for methods which saves images to
 * various file formats (but particularly the native SIF format)
 * 
 * TODO: Implement a thread-locking mechanism preventing RootPanel from
 * 	closing with System.close if there is a Save action going on.
 */
public class SaveEngine implements ActionListener, MWorkspaceObserver {

	private final Timer autosaveTimer = new Timer(10000, this);
	
	public SaveEngine( MasterControl master) {
		autosaveTimer.start();
		
		master.addWorkspaceObserver(this);
	}
	
	
	// Autosave functionality 
	@Override
	public void actionPerformed(ActionEvent evt) {
		long time = System.currentTimeMillis();
		for( AutoSaveWatcher watcher : watchers) {
			if( watcher.workspace.getFile() != null &&
				watcher.interval < (time - watcher.lastTime)/1000 &&
				watcher.undoCount < watcher.workspace.getUndoEngine().getMetronome() - watcher.lastUndoSpot) 
			{
				watcher.lastTime = time;
				(new Thread(new Runnable() {
					@Override
					public void run() {
						MDebug.log("Saving Backup");
						saveWorkspace( watcher.workspace, new File(  watcher.workspace.getFile().getAbsolutePath() + "~"), false);
						MDebug.log("Finished");
					}
				})).start();
				watcher.lastTime = time;
				watcher.lastUndoSpot= watcher.workspace.getUndoEngine().getMetronome();
			}
		}
	}
	
	// Behavior might warrant a Map<Workspace,ASW>, but the syntax would
	//	be dirtier and there is no real performanc threat here.
	private final List<AutoSaveWatcher> watchers = new ArrayList<>();
	
	private class AutoSaveWatcher {
		final ImageWorkspace workspace;
		int interval;
		int undoCount;
		long lastTime;
		int lastUndoSpot;
		
		AutoSaveWatcher( ImageWorkspace workspace, int interval, int undoCount) {
			this.workspace = workspace;
			this.interval = interval;
			this.undoCount = undoCount;

			this.lastTime = System.currentTimeMillis();
			this.lastUndoSpot = workspace.getUndoEngine().getMetronome();
		}
	}
	
	/**
	 * If interval is positive then it will auto-save the workspace every 
	 * <code>interval</code> seconds if any changes have been made.  If
	 * <code>undoCount</cod> is positive, it will auto-save every that many
	 * undo's.  If both are set it'll reset only when both conditions are
	 * true.
	 */
	public void triggerAutosave( ImageWorkspace workspace, int interval, int undoCount ) {
		if( interval <= 0 && undoCount <= 0) return;
		
		for( AutoSaveWatcher watcher : watchers) {
			if( watcher.workspace == workspace) {
				watcher.interval = interval;
				watcher.undoCount = undoCount;
				return;
			}
		}
		
		watchers.add(new AutoSaveWatcher(workspace, interval, undoCount));
	}
	public void untriggerAutosave( ImageWorkspace workspace) {
		Iterator<AutoSaveWatcher> it = watchers.iterator();
		while( it.hasNext()) {
			if( it.next().workspace == workspace) {
				it.remove();
			}
		}
	}
	
	public void removeAutosaved( ImageWorkspace workspace) {
/*		File f = workspace.getFile();
		if( f != null) {
			 f= new File(workspace.getFile().getAbsolutePath() + "~");
			 if( f.exists()) {
				 f.delete();
			 }
		}*/
	}
	

	/** Attempts to save the workspace to a SIF (native image format) file. */
	public void saveWorkspace( ImageWorkspace workspace, File file) {
		saveWorkspace( workspace, file, true);
	}
	public void saveWorkspace( ImageWorkspace workspace, File file, boolean track) {

		SaveHelper helper = new SaveHelper(workspace);

		try {
			if( file.exists()) {
				file.delete();
			}
			file.createNewFile();
			
			helper.ra = new RandomAccessFile(file, "rw");
	
			
			
			// [4] Header
			helper.ra.write( SaveLoadUtil.getHeader());
			// [4]Version
			helper.ra.writeInt(SaveLoadUtil.VERSION);
			
			// [2] Width, [2] Height
			helper.ra.writeShort(workspace.getWidth());
			helper.ra.writeShort(workspace.getHeight());

			// Save Various Chunks
			saveGroupTree( helper);
			saveImageData( helper);
			saveAnimationData( helper);
			
			helper.ra.close();
			
			if( track)
				workspace.fileSaved(file);
		}catch (UnsupportedEncodingException e) {
			MDebug.handleError(ErrorType.FILE, null, "UTF-8 Format Unsupported (somehow).");
		}catch( IOException e) {}
		finally {
			try {
				if( helper.ra != null)
					helper.ra.close();
			} catch (IOException e) {}
		}
		workspace.relinquishCache(helper.reservedCache);
	}
	
	/** Saves the GRPT chunk containing the Group Tree data */
	private void saveGroupTree( SaveHelper helper) 
		throws UnsupportedEncodingException, IOException
	{
		byte depth = 0;
		helper.ra.write( "GRPT".getBytes("UTF-8"));
		long startPointer = helper.ra.getFilePointer();
		helper.ra.writeInt(0);
		
		_sgt_rec( helper, helper.dupeRoot, depth);

		long endPointer = helper.ra.getFilePointer();
		if( endPointer - startPointer > Integer.MAX_VALUE )
			MDebug.handleError( ErrorType.OUT_OF_BOUNDS, null, "Image Data Too Big (>2GB).");
		helper.ra.seek(startPointer);
		helper.ra.writeInt( (int)(endPointer - startPointer - 4));
		
		helper.ra.seek(endPointer);
	}
	// Recursive should be fine since there's a depth limit of 255
	private void _sgt_rec( SaveHelper helper, GroupTree.Node node, byte depth) 
			throws UnsupportedEncodingException, IOException 
	{
		// [1] : Depth of Node in GroupTree
		helper.ra.writeByte( depth);
		
		// [4] : alpha
		helper.ra.writeFloat(node.getAlpha());
		
		// [2] : x offset [2] : y offset
		helper.ra.writeShort( node.getOffsetX());
		helper.ra.writeShort(node.getOffsetY());

		// [1] : bitmask
		int mask = 
				(node.isVisible() ? SaveLoadUtil.VISIBLE_MASK : 0) | 
				(node.isExpanded() ? SaveLoadUtil.EXPANDED_MASK : 0);

		helper.ra.writeByte( mask);
		
		// [n] : Null-terminated UTF-8 String for Layer name
		helper.ra.write( SaveLoadUtil.strToByteArrayUTF8( node.getName()));

		helper.nodeMap.put( node, helper.nmMet++);
		if( node instanceof GroupTree.GroupNode) {
			GroupTree.GroupNode gnode = (GroupTree.GroupNode) node;
			// [1] : Node Type ID
			helper.ra.write( SaveLoadUtil.NODE_GROUP);
			
			// Go through each of the Group Node's children recursively and save them
			for( GroupTree.Node child : node.getChildren()) {
				if( depth == 0xFF)
					MDebug.handleWarning( WarningType.STRUCTURAL, null, "Too many nested groups (255 limit).");
				else {
					_sgt_rec( helper, child, (byte) (depth+1));
				}
			}
		}
		else if( node instanceof GroupTree.LayerNode) {
			Layer layer = ((GroupTree.LayerNode) node).getLayer();
			
			
			if( layer instanceof SimpleLayer) {
				// [1] : Node Type ID
				helper.ra.write( SaveLoadUtil.NODE_SIMPLE_LAYER);
				
				ImageHandle data = ((SimpleLayer) layer).getData();
				// [4] : ID of ImageData linked to this LayerNode
				helper.ra.writeInt( data.getID());
			}
			if( layer instanceof RigLayer) {
				RigLayer rig = (RigLayer)layer;
				List<Part> parts = rig.getParts();
				
				
				// [1] : Node Type ID
				helper.ra.writeByte(SaveLoadUtil.NODE_RIG_LAYER);
				
				// [1] : Number or Parts
				helper.ra.writeByte( parts.size());
				
				// per Part:
				for(Part part : parts) {
					// [n]: null-terminated UTF-8 String
					helper.ra.write(SaveLoadUtil.strToByteArrayUTF8(part.getTypeName()));
					
					// [2] : OffsetX, [2] :OffsetY
					helper.ra.writeShort(part.getOffsetX());
					helper.ra.writeShort(part.getOffsetY());
					
					// [4] : Depth
					helper.ra.writeInt( part.getDepth());
					
					// [4] ImageHandle ID
					helper.ra.writeInt(part.getImageHandle().getID());
				}
			}
		}
		else {
			MDebug.handleWarning(WarningType.STRUCTURAL, null, "Unknown GroupTree Node type on saving.");
		}
	}
	
	/** Saves the IMGD chunk containing the Image data, each in PNG format. */
	private void saveImageData( SaveHelper helper) 
			throws UnsupportedEncodingException, IOException 
	{
		// [4] : "IMGD" tag
		helper.ra.write( "IMGD".getBytes("UTF-8"));
		long filePointer = helper.ra.getFilePointer();
		
		// [4] : Length of ImageData Chunk
		helper.ra.writeInt(0);
		
		ByteArrayOutputStream bos = new ByteArrayOutputStream();
		for( ImageHandle part : helper.reservedCache) {
			// (Foreach ImageData)
			ImageIO.write( part.deepAccess(), "png", bos);

			// [4] : Image ID
			helper.ra.writeInt( part.getID());
			// [4] : Size of Image Data
			helper.ra.writeInt( bos.size());
			// [x] : Image Data
			helper.ra.write(bos.toByteArray());
			bos.reset();
		}
		
		long filePointer2 = helper.ra.getFilePointer();
		helper.ra.seek(filePointer);
		
		if( filePointer2 - filePointer > Integer.MAX_VALUE )
			MDebug.handleError( ErrorType.OUT_OF_BOUNDS, null, "Image Data Too Big (>2GB).");
		helper.ra.writeInt( (int)(filePointer2 - filePointer-4));
		
		helper.ra.seek(filePointer2);
	}
	
	/** Saves the ANIM chunk containing the Animation structures */
	private void saveAnimationData( SaveHelper helper)
			throws UnsupportedEncodingException, IOException 
	{
		List<Animation> animations = helper.workspace.getAnimationManager().getAnimations();
		
		// [4] : "ANIM" tag
		helper.ra.write( "ANIM".getBytes("UTF-8"));
		
		long start = helper.ra.getFilePointer();
		// [4] : ChunkLength (placeholder for now)
		helper.ra.writeInt(0);
		
		for( Animation animation : animations) {
			// [n] : UTF8: Animation Name
			helper.ra.write(SaveLoadUtil.strToByteArrayUTF8(animation.getName()));
			
			if( animation instanceof FixedFrameAnimation) {
				// [1] : ID
				helper.ra.writeByte(SaveLoadUtil.ANIM_FIXED_FRAME);
				
				// [2] : Number of Layers
				List<AnimationLayer>layers = ((FixedFrameAnimation) animation).getLayers();
				helper.ra.writeShort( layers.size());
				
				for( AnimationLayer layer : layers) {
					// [4] : Group Node Bound to
					helper.ra.writeInt(helper.nodeMap.get(layer.getGroupLink()));
					
					// [2] : Number of Frames
					List<Frame> frames = layer.getFrames();
					helper.ra.writeShort(frames.size());
					
					for( Frame frame : frames) {
						// [1] : Marker
						helper.ra.writeByte(frame.getMarker().ordinal());
						
						// [2] : Length
						helper.ra.writeShort(frame.getLength());
						
						if( frame.getMarker() == Marker.FRAME) {
							// [4] : LayerID (corresponding to order it appears in the file)
							helper.ra.writeInt( helper.nodeMap.get(frame.getLayerNode()));
						}
					}
				}
				
			}
			else {
				helper.ra.writeByte(SaveLoadUtil.UNKNOWN);
			}
		}
		
		
		long end = helper.ra.getFilePointer();
		helper.ra.seek(start);
		if( end - start> Integer.MAX_VALUE )
			MDebug.handleError( ErrorType.OUT_OF_BOUNDS, null, "Image Data Too Big (>2GB).");
		helper.ra.writeInt( (int)(end - start-4));
	}
	
	private class SaveHelper {
		final Map<Node, Integer> nodeMap = new HashMap<>();
		int nmMet = 0;
		final List<ImageHandle> reservedCache;
		final GroupNode dupeRoot;
		final ImageWorkspace workspace;
		RandomAccessFile ra;
		
		
		SaveHelper( ImageWorkspace workspace) {
			this.workspace = workspace;
			reservedCache = workspace.reserveCache();
			dupeRoot =  workspace.getRootNode();
		}
	}


	// :::: MWorkspaceObserver
	@Override public void currentWorkspaceChanged(ImageWorkspace selected, ImageWorkspace previous) {}
	@Override public void newWorkspace(ImageWorkspace newWorkspace) {}
	@Override public void removeWorkspace(ImageWorkspace workspace) {
		untriggerAutosave( workspace);
	}


}

File: src/main/kotlin/spirite/base/file/save/SaveEngine.kt
package spirite.base.file.save

import rb.vectrix.linear.Vec2i
import rb.vectrix.mathUtil.d
import sgui.swing.hybrid.Hybrid
import sgui.swing.hybrid.MDebug
import sgui.swing.hybrid.MDebug.ErrorType
import sgui.swing.hybrid.MDebug.WarningType.STRUCTURAL
import sgui.swing.hybrid.MDebug.WarningType.UNSUPPORTED
import spirite.base.file.SaveLoadUtil
import spirite.base.file.SaveLoadUtil.FFALAYER_CASCADING
import spirite.base.file.SaveLoadUtil.FFALAYER_GROUPLINKED
import spirite.base.file.SaveLoadUtil.FFALAYER_LEXICAL
import spirite.base.file.writeFloatArray
import spirite.base.file.writeUFT8NT
import spirite.base.imageData.IImageWorkspace
import spirite.base.imageData.animation.Animation
import spirite.base.imageData.animation.ffa.FFALayer.FFAFrame
import spirite.base.imageData.animation.ffa.FfaFrameStructure.Marker.*
import spirite.base.imageData.animation.ffa.FfaLayerCascading
import spirite.base.imageData.animation.ffa.FfaLayerGroupLinked
import spirite.base.imageData.animation.ffa.FfaLayerLexical
import spirite.base.imageData.animation.ffa.FixedFrameAnimation
import spirite.base.imageData.animationSpaces.FFASpace.FFAAnimationSpace
import spirite.base.imageData.groupTree.GroupTree.*
import spirite.base.imageData.layers.SimpleLayer
import spirite.base.imageData.layers.sprite.SpriteLayer
import spirite.base.imageData.mediums.magLev.MaglevFill
import spirite.base.imageData.mediums.magLev.MaglevStroke
import spirite.base.telemetry.TelemetryEvent
import java.io.File
import java.io.RandomAccessFile
import kotlin.math.min

class SaveContext(
        val workspace: IImageWorkspace,
        val ra: RandomAccessFile)
{
    val nodeMap = mutableMapOf<Node, Int>()
    val animationMap = mutableMapOf<Animation,Int>()
    val telemetry = TelemetryEvent()

    val root = workspace.groupTree.root
    val floatingData = workspace.mediumRepository.dataList
            .mapNotNull { workspace.mediumRepository.floatData(it) { medium -> MediumPreparer.prepare(medium) } }


    inline fun writeChunk(tag: String, crossinline writer : (RandomAccessFile)->Unit) {
        telemetry.runAction("Write Chunk: $tag") {
            if (tag.length != 4) {
                // Perhaps overkill, but this really should be a hard truth
                MDebug.handleError(ErrorType.FATAL, "Chunk types must be 4-length")
            }

            // [4] : Chunk Tag
            ra.write(tag.toByteArray(charset("UTF-8")))

            val start = ra.filePointer
            // [4] : ChunkLength (placeholder for now
            ra.writeInt(0)

            writer.invoke(ra)

            val end = ra.filePointer
            ra.seek(start)
            if (end - start > Integer.MAX_VALUE)
                MDebug.handleError(ErrorType.OUT_OF_BOUNDS, "Image Data Too Big (>2GB).")
            ra.writeInt((end - start - 4).toInt())
            ra.seek(end)

            telemetry.mark("size", (end - start).d)
        }
    }
}

object SaveEngine {

    fun saveWorkspace( file : File, workspace: IImageWorkspace) {
        if( file.exists())
        {
            val backup = File(file.absolutePath+"~")
            if( backup.exists())
                backup.delete()
            val canWrite = file.canWrite()
            val x = file.renameTo(backup)
            val deleted = file.delete()
        }

        file.createNewFile()

        val ra = RandomAccessFile(file, "rw")
        val context = SaveContext(workspace, ra)

        saveHeader(context)
        saveGroupTree(context)
        saveImageData(context)
        if( workspace.animationManager.animations.any())
            saveAnimationData(context)
        if( workspace.animationSpaceManager.animationSpaces.any())
            saveAnimationSpaceChunk(context)
        savePaletteData(context)
        PaletteMapSaver.savePaletteData(context)

        ra.close()
    }

    private fun saveHeader(context: SaveContext) {
        val ra = context.ra
        val workspace = context.workspace

        // [4] Header
        ra.write(SaveLoadUtil.header)
        // [4] Version
        ra.writeInt(SaveLoadUtil.version)

        // [2] Width, [2] Height
        ra.writeShort(workspace.width)
        ra.writeShort(workspace.height)
    }

    /** GRPT chunk saving the structure of the PrimaryGroupTree and all Nodes/Layers within it */
    private fun saveGroupTree( context: SaveContext)
    {
        context.writeChunk("GRPT") {ra->
            val depth = 0
            var met = 0

            fun buildReferences( node: Node, depth: Int) {
                // Fills out a map from nodes to an int identifier (constructed here) for use in any nodes that
                //  reference other nodes (such as ReferenceLayers)
                context.nodeMap.put( node, met++)
                (node as? GroupNode)?.children?.forEach { buildReferences(it, depth+1) }
            }
            fun writeNode( node: Node, depth: Int) {
                // [1] : Depth of GroupNode in GroupTree
                ra.writeByte(depth)

                // [4] : alpha
                ra.writeFloat( node.alpha)

                // [2] : xi offset, [2] : yi offset
                ra.writeShort(node.x)
                ra.writeShort(node.y)

                // [1] : bitmask
                ra.writeByte(
                        if( node.visible) SaveLoadUtil.VisibleMask else 0 or
                        if( node.expanded) SaveLoadUtil.ExpandedMask else 0 or
                        if( node is GroupNode && node.flatenned) SaveLoadUtil.FlattenedMask else 0)

                // [n], UTF8 : Layer Name
                ra.write(SaveLoadUtil.strToByteArrayUTF8(node.name))

                when( node) {
                    is GroupNode -> {
                        // [1] : NodeTypeId
                        ra.writeByte(SaveLoadUtil.NODE_GROUP)

                        // Go through each of the Group GroupNode's children recursively and save them
                        when( depth) {
                            0xFF -> if( node.children.any()) MDebug.handleWarning(STRUCTURAL, "Too many nested groups (255 limit), some nodes ignored.")
                            else -> node.children.forEach { writeNode(it, depth+1) }
                        }
                    }
                    is LayerNode -> {
                        val layer = node.layer
                        when( layer) {
                            is SimpleLayer -> {
                                // [1] : NodeTypeId
                                ra.writeByte(SaveLoadUtil.NODE_SIMPLE_LAYER)

                                // [4] : MediumId of Medium attatched to the SimpleLayer
                                ra.writeInt( layer.medium.id)
                            }
                            is SpriteLayer -> {
                                ra.writeByte(SaveLoadUtil.NODE_SPRITE_LAYER) // [1] : NodeTypeId
                                ra.writeByte(layer.type.permanentCode)  // [1] : Sprite LAyer Medium Type

                                val parts = layer.parts.toList()

                                // [1] : Number of parts
                                ra.writeByte( parts.size)

                                parts.forEach { part ->
                                    // Per Part:
                                    ra.write(SaveLoadUtil.strToByteArrayUTF8(part.partName)) // n : PartTypeName
                                    ra.writeFloat(part.transX)  // 4 : TranslationX
                                    ra.writeFloat(part.transY)  // 4 : TranslationY
                                    ra.writeFloat(part.scaleX)  // 4 : ScaleX
                                    ra.writeFloat(part.scaleY)  // 4 : ScaleY
                                    ra.writeFloat(part.rot)     // 4 : rotation
                                    ra.writeInt(part.depth)     // 4 : draw depth
                                    ra.writeInt(part.handle.id) // 4 : MediumId
                                    ra.writeFloat(part.alpha)   // 4 : Alpha
                                }
                            }
                            //is ReferenceLayer -> {}
                            //is PuppetLayer -> {}
                        }
                    }
                }

            }

            buildReferences(context.workspace.groupTree.root, depth)
            writeNode(context.root, 0)
        }
    }

    /** PLTT chunk containing the Palettes saved in the workspace */
    private fun savePaletteData( context: SaveContext) {
        context.writeChunk("PLTT") {ra ->
            context.workspace.paletteSet.palettes.forEach {
                // [n], UTF8 : Palette Name
                ra.write(SaveLoadUtil.strToByteArrayUTF8(it.name))

                val raw = it.compress()
                ra.writeShort(raw.size) // [2] Palette Data Size
                ra.write(raw)           // [n] Compressed Palette Data
            }
        }
    }

    /** IMGD chunk containing all Medium Data, with images saved in PNG Format*/
    private fun saveImageData( context: SaveContext) {
        context.writeChunk("IMGD") {ra->
            context.floatingData.forEach { floatingMedium ->
                // [4] : Medium Handle Id
                ra.writeInt( floatingMedium.id)

                val prepared = floatingMedium.condensed
                when( prepared) {
                    is PreparedFlatMedium -> {
                        ra.writeByte(SaveLoadUtil.MEDIUM_PLAIN) // [1] : Medium Type

                        val byteArray = Hybrid.imageIO.writePNG(prepared.image)
                        ra.writeInt( byteArray.size)    // [4] : Size of Image Data
                        ra.write(byteArray)             // [n] : Image Data
                    }
                    is PreparedDynamicMedium -> {
                        ra.writeByte(SaveLoadUtil.MEDIUM_DYNAMIC) // [1] : Medium Type
                        ra.writeShort( prepared.offsetX)        // [2] : Dynamic AnimationCommand Offset
                        ra.writeShort( prepared.offsetY)        // [2] : Dynamic Y Offset

                        val byteArray = prepared.image?.run { Hybrid.imageIO.writePNG(this)}
                        ra.writeInt( byteArray?.size ?: 0)  // [4] : Size of Image Data (note: May be 0)
                        byteArray?.run { ra.write(this)}    // [n] : Image Data
                    }
                    is PreparedMaglevMedium -> {
                        ra.writeByte(SaveLoadUtil.MEDIUM_MAGLEV)    // [1] : Medium Type
                        ra.writeShort(prepared.things.size)     // [2] : number of things
                        prepared.things.forEach {thing ->
                            when(thing) {
                                is MaglevStroke -> {
                                    ra.writeByte(SaveLoadUtil.MAGLEV_THING_STROKE)  // [1] : Thing type
                                    ra.writeInt(thing.params.color.argb32)             // [4] : Color
                                    ra.writeByte(thing.params.method.fileId)            // [1] : Method
                                    ra.writeFloat(thing.params.width)                  // [4] : Stroke Width
                                    ra.writeByte(thing.params.mode.fileId)        // [1] : Mode
                                    ra.writeInt(thing.drawPoints.length)     // [4] : Num Vertices

                                    ra.writeFloatArray(thing.drawPoints.x)
                                    ra.writeFloatArray(thing.drawPoints.y)
                                    ra.writeFloatArray(thing.drawPoints.w)
                                }
                                is MaglevFill -> {
                                    ra.writeByte(SaveLoadUtil.MAGLEV_THING_FILL)    // [1] : ThingType
                                    ra.writeInt(thing.color.argb32) // [4] : Color
                                    ra.writeByte(thing.mode.fileId) // [1] : FillMode

                                    val numSegs = min(65535, thing.segments.size)
                                    ra.writeShort(numSegs)  // [2] : Num Segments
                                    thing.segments.forEach { seg ->
                                        ra.writeInt(seg.strokeId)   // [4]: StrokeId
                                        ra.writeInt(seg.start)  // [4] : Start
                                        ra.writeInt(seg.end)    // [4]
                                    }
                                }
                            }
                        }
                        val byteArray = prepared.image?.run { Hybrid.imageIO.writePNG(this)}
                        ra.writeInt( byteArray?.size ?: 0)  // [4] : Size of Image Data
                        ra.writeShort( prepared.offsetX)        // [2] : Dynamic AnimationCommand Offset
                        ra.writeShort( prepared.offsetY)        // [2] : Dynamic Y Offset
                        byteArray?.run { ra.write(this)}    // [n] : Image Data
                    }
                }
            }
        }
    }

    private fun saveAnimationSpaceChunk(context: SaveContext) {
        val animMap = context.animationMap

        context.writeChunk("ANSP") {ra ->
            context.workspace.animationSpaceManager.animationSpaces.forEach { space ->
                ra.writeUFT8NT(space.name)

                when( space) {
                    is FFAAnimationSpace -> {
                        ra.writeByte(SaveLoadUtil.ANIMSPACE_FFA) // [1] : Type

                        val animations = space.animationStructs.toList()
                        val links = space.links.toList()

                        ra.writeShort(animations.size)  // [2] : Number of Animations
                        animations.forEach { struct ->
                            val anim = struct.animation
                            ra.writeInt(animMap[anim] ?: -1)    // 4: AnimationId
                            val onEnd = struct.onEndLink
                            val onEndLink =  if( onEnd == null) -1 else animMap[onEnd.first] ?: -1

                            ra.writeInt(onEndLink) // 4: AnimationId of on-end groupLink
                            if(onEndLink != -1)
                                ra.writeInt(onEnd!!.second)    // 4: on-end Frame

                            val logSpace = space.stateView.logicalSpace[anim] ?: Vec2i.Zero
                            ra.writeShort(logSpace.xi)    // 2: Logical AnimationCommand
                            ra.writeShort(logSpace.yi)    // 2: Logical Y
                        }

                        ra.writeShort(links.size)   // [2] : Number of Links
                        links.forEach {
                            ra.writeInt(animMap[it.origin] ?: -1)
                            ra.writeInt(it.originFrame)
                            ra.writeInt(animMap[it.destination] ?: -1)
                            ra.writeInt(it.destinationFrame)
                        }
                    }
                    else -> {
                        ra.writeByte(0)
                        MDebug.handleWarning(UNSUPPORTED, "Do not know how to save Animation Space: $space.  Skipping it")
                    }
                }
            }
        }
    }

    private const val MaxFFALayers = Short.MAX_VALUE.toInt()
    private const val MaxFFALayerFrames = Short.MAX_VALUE.toInt()

    /** ANIM chunk containing all Animation Data */
    private fun saveAnimationData( context: SaveContext) {
        context.writeChunk("ANIM") {ra ->
            var met = 0

            context.workspace.animationManager.animations.forEach { anim ->
                context.animationMap[anim] = met++

                when( anim) {
                    is FixedFrameAnimation -> {
                        ra.write(SaveLoadUtil.strToByteArrayUTF8(anim.name))    // [n] Anim name
                        ra.writeFloat(anim.state.speed)                         // [4] : Anim Speed
                        ra.writeShort(anim.state.zoom)                          // [2] : Anim Zoom
                        ra.writeByte(SaveLoadUtil.ANIM_FFA)                     // [1] : Anim TypeId

                        if(anim.layers.size > MaxFFALayers) MDebug.handleWarning(UNSUPPORTED, "Too many Animation layers (num: ${anim.layers.size} max: $MaxFFALayers), taking only the first N")

                        val writtenExplicits = hashSetOf<Node>()

                        ra.writeShort( min(anim.layers.size, MaxFFALayers))  // [2] : Number of layers
                        for (layer in anim.layers.asSequence().take(MaxFFALayers)){
                            ra.writeUFT8NT(layer.name)  // [n] : Layer Name
                            ra.writeByte( if(layer.asynchronous) 1 else 0)  // [1] : IsAsynchronous
                            when(layer) {
                                is FfaLayerGroupLinked -> {
                                    ra.writeByte(FFALAYER_GROUPLINKED)  // [1] : Layer TypeId

                                    ra.writeInt(context.nodeMap[layer.groupLink] ?: -1)    // [4] : NodeId of GroupNode Bount
                                    ra.writeByte(if(layer.includeSubtrees) 1 else 0)    // [1] : 0 bit : whether or not subgroups are linked

                                    if( layer.frames.size > MaxFFALayerFrames) MDebug.handleWarning(UNSUPPORTED, "Too many Frames in a layer (max: ${Short.MAX_VALUE}, only writing first N)")

                                    ra.writeShort(min(layer.frames.size, MaxFFALayerFrames)) // [2] : Number of Frames
                                    for( frame in layer.frames.asSequence().take(MaxFFALayerFrames)) {
                                        val type = when((frame as FFAFrame).marker) {
                                            GAP -> SaveLoadUtil.FFAFRAME_GAP
                                            START_LOCAL_LOOP -> SaveLoadUtil.FFAFRAME_STARTOFLOOP
                                            FRAME -> SaveLoadUtil.FFAFRAME_FRAME
                                            END_LOCAL_LOOP -> null
                                        } ?: continue
                                        ra.writeByte(type)
                                        ra.writeInt(context.nodeMap[frame.structure.node] ?: -1 ) // [4] : NodeId
                                        ra.writeShort(frame.length)    // [2]: Length
                                    }
                                }
                                is FfaLayerLexical -> {
                                    ra.writeByte(FFALAYER_LEXICAL)  // [1] : Layer TypeId
                                    ra.writeInt(context.nodeMap[layer.groupLink] ?: -1) // [4] NodeIf of GroupNode
                                    ra.writeUFT8NT(layer.lexicon)   // [n] : Lexicon

                                    if( writtenExplicits.contains(layer.groupLink)) {
                                        ra.writeByte(0) // [1] : No Explicits to write
                                    }
                                    else {
                                        ra.writeByte(min(255, layer.sharedExplicitMap.size))   // [1] : Num Explicits
                                        for( explicit in layer.sharedExplicitMap.asSequence().take(255)) {
                                            ra.writeByte(explicit.key.toByte().toInt()) // [1] : Char mapping
                                            ra.writeInt(context.nodeMap[explicit.value] ?: -1) // [4] : NodeId
                                        }
                                    }
                                }
                                is FfaLayerCascading -> {
                                    ra.writeByte(FFALAYER_CASCADING)    // [1] : Layer TypeId
                                    ra.writeInt(context.nodeMap[layer.groupLink] ?: -1) // [4] NodeId of GroupNode
                                    ra.writeUFT8NT(layer.lexicon ?: "") // [n] : Lexicon

                                    val subinfos = layer.sublayerInfo.values.toList()
                                    ra.writeByte(min(255, subinfos.size)) // [1] : Group Subinfos
                                    subinfos.forEach {
                                        ra.writeInt(context.nodeMap[it.group] ?: -1)    // [4] NodeId
                                        ra.writeShort(it.primaryLen)                        // [2] PrimaryLen
                                        ra.writeByte(it.lexicalKey.toByte().toInt())        // [1] Lexical Key
                                        ra.writeUFT8NT(it.lexicon ?: "")
                                    }
                                }
                            }

                        }
                    }
                    else -> MDebug.handleWarning(UNSUPPORTED, "Do not know how to save Animation: $anim.  Skipping it")
                }
            }
        }

    }
}